### PR TITLE
Joshuabg/crypto feature

### PR DESCRIPTION
**Pull request by adithyasolai**
Best implementation of Crypto Feature for Assignment 2.

### CODE REVIEW

File: src/test/java/net/testudobank/helpers/MvcControllerIntegTestHelpers.java [hunk 1]
```diff
@@ -70,6 +79,19 @@ public static void checkOverdraftLog(Map<String,Object> overdraftLog, LocalDateT
     System.out.println("Timestamp stored in OverdraftLogs table for the Repayment: " + overdraftLogTimestamp);
   }
 
+  // Verifies that a single crypto log in the CryptoHistory table matches the expected customerID, timestamp, action, cryptocurrency, and amount
+  public static void checkCryptoLog(Map<String,Object> cryptoLog, LocalDateTime timeWhenRequestSent, String expectedCustomerID, String expectedAction, String expectedCryptoName, double expectedCryptoAmount) {
+    assertEquals(expectedCustomerID, cryptoLog.get("CustomerID"));
+    assertEquals(expectedAction, cryptoLog.get("Action"));
+    assertEquals(expectedCryptoAmount, ((BigDecimal) cryptoLog.get("CryptoAmount")).doubleValue());
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/helpers/MvcControllerIntegTestHelpers.java:86]: I have a small concern that the conversion from `BigDecimal` to `double` can sometimes make it not equal the `double expectedCryptoAmount`, but this is probably negligible when considering that this is in a controlled integ test environment, and that we will be using simple CryptoAmount values.

File: src/main/java/net/testudobank/MvcController.java [hunk 2]
```diff
@@ -619,23 +612,167 @@ public String submitTransfer(@ModelAttribute("user") User sender) {
...
+
+    // check if user has required crypto balance
+    // TODO: comparing doubles like this is probably not a good idea
+    if (cryptoBalance.get() < cryptoAmountToSell) {
+      return "welcome";
+    }
+
+    double cashValueOfCrypto = cryptoPriceClient.getCurrentEthValue() * cryptoAmountToSell;
+
+    String currentTime = SQL_DATETIME_FORMATTER.format(new java.util.Date());
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/MvcController.java:757]: Nitpick: Probably a bit better for this to go after the 2 `user.set()` lines below and directly before `submitDeposit()`. Not a big deal.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 3]
```diff
@@ -48,18 +54,19 @@
     .withPassword("db_password")
     .withDatabaseName("testudo_bank");
 
-  
+
   private static MvcController controller;
   private static JdbcTemplate jdbcTemplate;
   private static DatabaseDelegate dbDelegate;
+  private static CryptoPriceClient cryptoPriceClient = Mockito.mock(CryptoPriceClient.class);
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:61]: What is the point of mocking this? Why not just use a live `CryptoPriceClient`?

Comment (adithyasolai) [reply src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:61]: Read a bit further. Mocking this makes sense. You are allowing the tester to input the Ethereum Price.

Doesn't this kind of defeat the purpose of an integration/system test if you are giving so much control over things like what an external Crypto Price API returns?

File: src/test/java/net/testudobank/helpers/MvcControllerIntegTestHelpers.java (hunk 1 above)
Comment (joshg4096) [reply src/test/java/net/testudobank/helpers/MvcControllerIntegTestHelpers.java:86]: Alternatively, I thought of changing everything in the application involving crypto from double to BigDecimal. This is a substantial change though

**Pull request merged**